Web Extensions: Parsing for Create() Bookmark function

akutira-umich opened the pull request:
#### 7c3bccb28c9d00d450851dd40dcc5a3e4554f8ce
<pre>
Web Extensions: Parsing for Create() Bookmark function
<a href="https://rdar.apple.com/154231013">rdar://154231013</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294932">https://bugs.webkit.org/show_bug.cgi?id=294932</a>

Reviewed by Timothy Hatcher.

Parses the create function using a CreateDetails Object which takes in the parameters
for the dictionary. For now this info from the dictionary is put into MockBookmarkNodes.
Also makes sure default type of a bookmark create request is assigned properly and
dateAdded is a parameter you can set only for testing purposes.
Created tests for these as well.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::toWebAPI):
(WebKit::toTypeImpl):
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getRecent):
(WebKit::WebExtensionAPIBookmarks::createDictionaryFromNode):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICreateParse)):

Canonical link: <a href="https://commits.webkit.org/296824@main">https://commits.webkit.org/296824@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2092a9f6dcff138b06cebf4b7d80ea972f2f99a0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109692 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29350 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19779 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115713 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59926 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111655 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30028 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37938 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83355 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112640 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23924 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98781 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63814 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23306 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16926 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59507 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93299 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16963 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118505 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36731 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27203 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92364 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37104 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95040 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92185 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37143 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14887 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/32559 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17703 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36625 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42096 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36286 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39628 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37995 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -39,26 +39,65 @@
+        return folderKey;
+    }
+
+    if (inputURL && [inputURL length] > 0)
```
A few nits here:

- If `inputURL` is nil, then `[inputURL length]` is 0, so you don't need both of these checks.
- Our style is `inputURL.length`

So this can just become:
`if (inputURL.length)
    return bookmarkKey;
`

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -39,26 +39,65 @@
 static NSString * const urlKey = @"url";
 static NSString * const titleKey = @"title";
+static NSString * const indexKey = @"index";
+static NSString * const parentidKey = @"parentId";
```
Super nit: `parentIdKey` or `parentIDKey`

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -67,15 +106,33 @@
 
     m_mockBookmarks.append(WTFMove(newNode));
-    callback->call(createDictionaryFromNode(newNode));
+    callback->call(createDictionaryFromNode(m_mockBookmarks[m_mockBookmarks.size() - 1]));
```
Nit: `m_mockBookmarks.last()`

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -67,15 +106,33 @@
 
-    newNode.dateAdded = WallTime::fromRawSeconds(objectForKey<NSDate>(bookmark, dateAddedKey).timeIntervalSince1970);
+    double msSinceEpoch = dynamic_objc_cast<NSNumber>(bookmark[dateAddedKey]).doubleValue;
+    newNode.dateAdded = WTF::WallTime::fromRawSeconds(msSinceEpoch / 1000.0);
```
```suggestion
    newNode.dateAdded = WTF::WallTime::fromRawSeconds(Seconds::fromMilliseconds(msSinceEpoch).value());
```

AriYoung00 replied:
Actually, is there a reason you don't want to keep the previous approach? I.E.:
`newNode.dateAdded = WallTime::fromRawSeconds(objectForKey<NSDate>(bookmark, dateAddedKey).timeIntervalSince1970);`

Unless there are issues, I think that would be the best way to do things here.

akutira-umich replied:
Yea the issue was that it wasn't allowing dateAdded to be set. It would force the value to be 0. This is because of the NSDate type where dateAdded is an NSNumber.

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -67,15 +106,33 @@
     if (!validateDictionary(bookmark, @"bookmark", requiredKeys, types, outExceptionString))
         return;
 
+    parsedDetails.index = [static_cast<NSNumber*>(bookmark[indexKey]) intValue];
```
```suggestion
    parsedDetails.index = objectForKey<NSNumber>(bookmark, indexKey).intValue
```
And similarly below

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -35,14 +35,32 @@ namespace WebKit {
+
+    struct CreateDetails {
+        std::optional<int> index;
+        RetainPtr<NSString> parentId;
```
```suggestion
        String parentId;
```
And similarly below

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -39,26 +39,65 @@
+
+    return tempNode;
+}
+NSString* WebExtensionAPIBookmarks::stringFromBookmarkTreeNodeType(WebExtensionAPIBookmarks::BookmarkTreeNodeType type, NSString* inputURL)
```
```suggestion
static NSString *toWebAPI(WebExtensionAPIBookmarks::BookmarkTreeNodeType type, NSString *inputURL)
```

akutira-umich replied:
Is there a way to use MockBookmarkNode's if the function is no longer in the WebExtensionAPIBookmarks class?

xeenon replied:
Yes, `WebExtensionAPIBookmarks::MockBookmarkNode`. Might need to make it `public`.

akutira-umich replied:
Ahh ok, wasn't sure if I wanted to make it public but ill do that

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -39,26 +39,65 @@
     };
+
+    return tempNode;
+}
```
```suggestion
}

```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -39,26 +39,65 @@
+
+    return folderKey;
+}
+std::optional<WebExtensionAPIBookmarks::BookmarkTreeNodeType> WebExtensionAPIBookmarks::bookmarkTreeNodeTypeFromString(NSString* typeString)
```
```suggestion
}

static std::optional<WebExtensionAPIBookmarks::BookmarkTreeNodeType> toImpl(NSString *typeString)
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -67,15 +106,33 @@
+    parsedDetails.parentId = objectForKey<NSString>(bookmark, parentIdKey);
+    parsedDetails.title = objectForKey<NSString>(bookmark, titleKey);
+    parsedDetails.url = objectForKey<NSString>(bookmark, urlKey);
+    if (bookmark[typeKey]) {
```
```suggestion
    parsedDetails.url = objectForKey<NSString>(bookmark, urlKey);

    if (bookmark[typeKey]) {
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -67,15 +106,33 @@
+            *outExceptionString = toErrorString(nullString(), @"type", @"type must be bookmark or folder").createNSString().autorelease();
+            return;
+        }
+        parsedDetails.type = parsedType;
```
```suggestion
        }

        parsedDetails.type = parsedType;
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -67,15 +106,33 @@
+    if (bookmark[typeKey]) {
+        std::optional<WebExtensionAPIBookmarks::BookmarkTreeNodeType> parsedType = WebExtensionAPIBookmarks::bookmarkTreeNodeTypeFromString(bookmark[typeKey]);
+        if (!parsedType) {
+            *outExceptionString = toErrorString(nullString(), @"type", @"type must be bookmark or folder").createNSString().autorelease();
```
```suggestion
            *outExceptionString = toErrorString(nullString(), typeKey, @"it must specify either 'bookmark' or 'folder'").createNSString().autorelease();
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -35,14 +35,32 @@ namespace WebKit {
     };
 
+    std::optional<BookmarkTreeNodeType> bookmarkTreeNodeTypeFromString(NSString* typeString);
+    NSString* stringFromBookmarkTreeNodeType(BookmarkTreeNodeType type, NSString* inputURL);
```
These are not needed now.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -35,14 +35,32 @@ namespace WebKit {
+        String parentId;
         String title;
         String url;
+        int index;
```
```suggestion
        size_t index;
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -39,26 +39,65 @@
 static NSString * const bookmarkKey = @"bookmark";
+static NSString * const folderKey = @"folder";
 
 NSDictionary *WebExtensionAPIBookmarks::createDictionaryFromNode(const MockBookmarkNode& node)
```
This should also be static.
```suggestion
static NSDictionary *toWebAPI(const MockBookmarkNode& node)
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -67,15 +106,33 @@
+    newNode.parentId = bookmark[parentIdKey];
+    newNode.index = parsedDetails.index.value();
+    if (parsedDetails.type)
+        newNode.type = parsedDetails.type.value();
```
I think `newNode.type` will have a default value, even if not assigned explicitly here. If that's intentional behavior, I would make it explicit with:
```suggestion
    newNode.type = parsedDetails.type.value_or(/* default BookmarkTreeNodeType goes here */);
```
If that's not intentional behavior, then `MockBookmarkNode::type` should also be `optional`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -39,26 +39,67 @@
+
+static NSDictionary *toWebAPI(const WebExtensionAPIBookmarks::MockBookmarkNode& node)
+{
+    NSString* typeString = toWebAPI(node.type, node.url.createNSString().get());
```
```suggestion
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -39,26 +39,67 @@
+        indexKey: @(node.index),
         dateAddedKey: @(node.dateAdded.secondsSinceEpoch().milliseconds()),
-        typeKey: bookmarkKey
+        typeKey: typeString
```
```suggestion
        typeKey: toWebAPI(node.type, node.url.createNSString().get());
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -67,15 +108,32 @@
     if (!validateDictionary(bookmark, @"bookmark", requiredKeys, types, outExceptionString))
```
```suggestion
    if (!validateDictionary(bookmark, @"bookmark", nil, types, outExceptionString))
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -39,26 +39,67 @@
-    static NSDictionary<NSString *, id> *types= @{
+    WebExtensionAPIBookmarks::CreateDetails parsedDetails;
+
+    NSArray *requiredKeys = @[];
```
```suggestion
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -67,15 +108,32 @@
     if (!validateDictionary(bookmark, @"bookmark", requiredKeys, types, outExceptionString))
         return;
 
+    parsedDetails.index = objectForKey<NSNumber>(bookmark, indexKey).intValue;
```
```suggestion
    parsedDetails.index = objectForKey<NSNumber>(bookmark, indexKey).unsignedLongValue;
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -67,15 +108,32 @@
+    parsedDetails.url = objectForKey<NSString>(bookmark, urlKey);
+
+    if (bookmark[typeKey]) {
+        std::optional<WebExtensionAPIBookmarks::BookmarkTreeNodeType> parsedType = toImpl(dynamic_objc_cast<NSString>(bookmark[typeKey]));
```
```suggestion
        auto parsedType = toTypeImpl(dynamic_objc_cast<NSString>(bookmark[typeKey]));
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -39,26 +39,67 @@
+    return tempNode;
+}
+
+static std::optional<WebExtensionAPIBookmarks::BookmarkTreeNodeType> toImpl(NSString *typeString)
```
```suggestion
static std::optional<WebExtensionAPIBookmarks::BookmarkTreeNodeType> toTypeImpl(NSString *typeString)
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -67,15 +108,32 @@
+    if (bookmark[typeKey]) {
+        std::optional<WebExtensionAPIBookmarks::BookmarkTreeNodeType> parsedType = toImpl(dynamic_objc_cast<NSString>(bookmark[typeKey]));
+        if (!parsedType)
+            *outExceptionString = toErrorString(nullString(), typeKey, @"it must specify either 'bookmark' or 'folder'").createNSString().autorelease();
```
```suggestion
        if (!parsedType) {
            *outExceptionString = toErrorString(nullString(), typeKey, @"it must specify either 'bookmark' or 'folder'").createNSString().autorelease();
            return;
        }
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -34,20 +34,33 @@ namespace WebKit {
-private:
+
+public:
+    enum class BookmarkTreeNodeType {
```
```suggestion
    enum class BookmarkTreeNodeType : uint_8_t {
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -34,20 +34,33 @@ namespace WebKit {
+    };
+
+    struct CreateDetails {
+        std::optional<int> index;
```
```suggestion
        std::optional<size_t> index;
```

akutira-umich committed:
Web Extensions: Parsing for Create() Bookmark function
rdar://154231013
https://bugs.webkit.org/show_bug.cgi?id=294932

Reviewed by Timothy Hatcher.

Parses the create function using a CreateDetails Object which takes in the parameters
for the dictionary. For now this info from the dictionary is put into MockBookmarkNodes.
Also makes sure default type of a bookmark create request is assigned properly and
dateAdded is a parameter you can set only for testing purposes.
Created tests for these as well.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::toWebAPI):
(WebKit::toTypeImpl):
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getRecent):
(WebKit::WebExtensionAPIBookmarks::createDictionaryFromNode):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICreateParse)):

Canonical link: https://commits.webkit.org/296824@main
